sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala

```diff
@@ -453,45 +453,89 @@ case class RangeExec(range: org.apache.spark.sql.catalyst.plans.logical.Range)
     val localIdx = ctx.freshName("localIdx")
     val localEnd = ctx.freshName("localEnd")
     val range = ctx.freshName("range")
-    val shouldStop = if (parent.needStopCheck) {
-      s"if (shouldStop()) { $number = $value + ${step}L; return; }"
+
+    val processingLoop = if (parent.needStopCheck) {
+      // TODO (cloud-fan): do we really need to do the stop check within batch?
```

Comment:
This is the motivation of bringing the discussion at https://github.com/apache/spark/pull/10989#discussion_r221961271

If it's OK to not interrupt the loop and buffer result rows for join, I think it's also OK here.

Reply:
if we don't, then we would consume more rows than needed, don't we?

Reply:
we just buffer more rows in `BufferRowIterator.currentRows`, it's only about performance IIUC.

Reply:
mmmh, but `localIdx` would become `localEnd` then, right? So the UTs you added would fail, or am I missing something?

Reply:
I think that there is BroadcastHashJoin case doesn't mean it is generally ok to buffer more rows. If it is possible, we still should avoid it.

Reply:
> mmmh, but localIdx would become localEnd then, right? So the UTs you added would fail, or am I missing something?

You can pull my PR and try, the whole stage codegen is a little convoluted. `localIndex` is local and always starts with 0, `localEnd` is decided by the global `nextIndex`.